mitodl/ocw-studio
# Show confirmation dialog when data would be lost

#### Pre-Flight checklist


#### What are the relevant tickets?
Fixes #792
Fixes #793 

#### What's this PR do?
Show a confirmation modal in two circumstances:
 - if changes are made and the editing drawer is closed without saving changes for page, resource, or any of the other collection types
 - if changes are made to the menu or metadata page, or any other of the singleton pages, and the user navigates away from the page, or attempts to close the tab


#### How should this be manually tested?
This is tricky so be sure to go through all the steps to make sure nothing was overlooked. This should feel reasonably intuitive so please let me know if something feels awkward. Create or reuse a site with the ocw-course starter:

Adding a page:
 - Click "Add Pages" and watch the drawer open. Click away from the drawer. The drawer should close. The same should happen if you click the "x' on the upper right corner.
 - Click "Add Pages" and type some content. Click away from the drawer. You should see a confirmation modal saying "You have unsaved changes". 
   - Click the Cancel button and the confirmation modal should disappear. 
   - Click away from the drawer again and the confirmation modal should reappear. Click outside the confirmation modal. The confirmation modal should disappear and the drawer should stay open. 
   - Click away from the drawer again and the confirmation modal should reappear. Click "Discard changes". Both the confirmation modal and drawer should close, and you should not see any new pages added.
 - Click "Add Pages" again and type some content. Click "Save" and the drawer should close and a new page should appear in the list. 
   - Click "Metadata" on the side to test what happens when you redirect to another page. You should go to that page as usual.

Editing a page:
- Click on the title of the page you edited and a drawer should open with its contents. 
  - Type some extra content into a field. Click outside the drawer, and you should see a confirmation dialog. Click "Cancel" and the confirmation dialog should disappear but the drawer should stay open.
  - Close your tab. You should see a different confirmation message asking if you want to leave the page (the exact text depends on your browser). Click "Stay on page" or the equivalent. (If you click "Leave" it should close the browser tab.)
  - Click "Save" and the drawer should close. Click "Metadata" on the side and you should be redirected to that page.

Editing metadata:
 - Click on "Metadata" to go to that page. Type some content in a field, then click "Menu" in the sidebar. You should see a confirmation message asking if you want to discard changes. 
   - Click "ok". You should be redirected to the Menu page. Click on "Metadata" again and you should see the old metadata content without any changes that you made.
   - Enter some changes and click "Menu" again. It should show the confirmation dialog. Click "Cancel" and the confirmation dialog should disappear but you should remain on the same page.
   - Click "Save" at the bottom, then click on "Menu" again. You should be redirected to the menu page. You should be able to click between "Metadata" and "Menu" without seeing a confirmation dialog.

Menu:
 - Click "Add New" and a drawer should open. 
   - Note that there is no confirmation dialog for the drawer, it was challenging and tricky to implement two confirmation dialogs for the same page especially given the history API, so I left this out of the PR.
   - Type some content and click outside the drawer. It should close without a confirmation dialog.
   - Click "Add New" and add a link to the page you made. Click Save, the drawer should close and you should see the link listed.
   - Try to close the tab. You should see a confirmation dialog. Click "Cancel" and it should dismiss.
   - Click "Metadata". You should see a confirmation dialog. Click "ok" and you should be redirected to the metadata page. Click "Menu" and you should be back at the menu page, but without any changes you made.
 - Click "Add New" again and add an internal link to your page. Do this again and add an external link to some other page.
   - Click "Save". Then click "Metadata" to go to the metadata page. You should be redirected to that page. Click "Menu" and you should be redirected to the menu page, and your content should show up properly.
 - Refresh the browser. It should not show a confirmation dialog.
 - Reorder the two links somehow. Refresh the browser. It should show a confirmation dialog. 
 - Click Cancel, then click Save to save the new order for your links. Refresh the browser, and it should refresh without showing you a confirmation dialog.
 - Click on the gear of one of the links and make a change, then click "Save" in the drawer. The drawer should close. Click "Metadata" and you should see a confirmation dialog.

#### Screenshots
![Screenshot from 2021-11-10 14-48-41](https://user-images.githubusercontent.com/863262/141478331-6e2bbbcd-2dc2-46f0-8e5d-8a0d4206fdcc.png)
![Screenshot from 2021-11-12 15-54-32](https://user-images.githubusercontent.com/863262/141533552-5eeda10c-752c-4b47-9d91-5395a573f1a2.png)
